### ISSUE TITLE
[FR] Support Typescript

### ISSUE THREAD

**Issue by John0x** [strapi/strapi#630]
I believe that TypeScript support for both the admin panel and the API would be a great addition. Adding typescript support for the admin panel shouldn't be too much of a problem, since you're already using webpack for bundling, right?

**Comment by Aurelsicoko** [comment]
We are also thinking about using Flow... We will certainly make a survey to ensure the community is following us on this choice. We want to make sure everyone could contribute easily to the project like it's already the case.

**Comment by John0x** [comment]
@Aurelsicoko I'd be willing to take care of most of the typescript integration, if you would guarantee, that it will be merged.

I'll post it on the vote page.

**Comment by Aurelsicoko** [comment]
@John0x It could be awesome 👍

**Comment by oeddyo** [comment]
where's the vote? I'd love to vote for ts :D

**Comment by AlexKhare4** [comment]
+1

**Comment by erikvullings** [comment]
Unfortunately, there is still no option to vote for this feature. In the mean time, I've created a simple CLI, [strapi-to-typescript](https://www.npmjs.com/package/strapi-to-typescript), to generate typescript interfaces from the api/models/**.settings.json files.

```console
npm i -g strapi-to-typescript
sts [PATH\TO\STRAPI\API\FOLDER]
```

**Comment by TimMensch** [comment]
TypeScript is a best practice on Node today. I came to look for options for Loopback, and will be ignoring Strapi as a legacy project without TypeScript support.

There's no vote option, there's no mention of TypeScript anywhere, and there's no `@types/strapi` module. That pretty much kills Strapi as an option.

**Comment by derrickmehaffy** [comment]
@timmensch Your always welcome to develop it yourself and create a PR!

Strapi is an open source project and their team is quite small, naturally the community is what helps make a project!

Always have to think on the positive side :)

**Comment by NickBolles** [comment]
I would also contribute a lot to the typescript implementation. 

I think that having defined classes, interfaces and types for Services/Controllers/ORMs etc would be extremely beneficial to Strapi. 

@John0x  @erikvullings  @TimMensch have any of you made any progress on this? Should we get something going? 

Maybe we start with a small part and get the tooling in place and do things one by one

**Comment by TimMensch** [comment]
@NickBolles I was serious when I said I would be ignoring `strapi`. My comment above resulted in excuses and a request for someone else to do all the work -- which is non-trivial and would require the cooperation of the project. 

And then I get more downvotes than upvotes, so this community clearly hasn't gotten the memo that TypeScript is where we _all_ should be. As I said, it's a best practice today. Anyone who ignores that fact is going to fall behind, "small team" or no.

It's beyond just creating declarations, too. You'd need to create a generator for TypeScript signatures for each model. I've _done_ this before, but I don't really feel the need to join a community that's borderline hostile to TypeScript. Too much chance for conflict.

Loopback 2 will be 100% TypeScript (required! not optional! _They_ got the memo.); until Loopback 2 is release-quality, I'm using FeathersJS+Feathers-Sequelize+Sequelize-Typescript+Feathers-Swagger. Not perfect, but much better than trying to cobble together TypeScript support for a community that doesn't care.

If your project is risk-tolerant enough, you can use Loopback 2 today with full TypeScript support; they stilled called it "alpha" when I looked last, and my enterprise client wouldn't have been happy using an "alpha" quality product.

**Comment by abdonrd** [comment]
I think TypeScript could be a good thing. 🙂

**Comment by Aurelsicoko** [comment]
@TimMensch You're talking about Loopback 4, not LP2. We aren't reluctant to move to TS or Flow but there is a way to explain to a community why we should move to this kind of technology.

I'm 100% sure as well we should at least use Flow or move to TypeScript. However, even the Facebook team recommends using TS instead of Flow (https://twitter.com/sebmck/status/1056948355941584896). The only cons that I have against TypeScript is the ease to contribute to the project. It's far easier to contribute to a project without TS or Flow. However, the pros are stronger (type-checking, code confidence, etc) and we should definitely move to TS or Flow.

We are currently thinking about the implementation of TS for the next major release.

**Comment by TimMensch** [comment]
Sorry, haven't looked at Loopback for a while. You're right, it's 4.

Flow is just a non-starter. TypeScript has really won that battle, and it wasn't even a battle. As you point out, even Facebook isn't backing their own project any more. Tons of projects have built-in TypeScript support, or good `@types`, and the only ones I've seen in the wild with built-in Flow support have been Facebook projects. I'm sure they exist, but they're rare, and TypeScript is _ubiquitous_. 

Good to know that you're considering it.

On ease of contribution to a project: That will change over time. TypeScript keeps going up in popularity (for good reason). I think you'll find that average contribution _quality_ is higher too. 

And the _hard part_ about TypeScript is getting the project up and running, including all of the type dependencies and proper definitions and casting. Once you have that, the rest is actually pretty easy. Contributing, when you see examples, isn't a whole lot harder than contributing straight JavaScript.

**Comment by NickBolles** [comment]
I would actually argue that contributing to a typescript project is much easier after you understand the absolute basics of typescript (maybe a couple hours).

It improves the visibility and discovarability of the APIs and data structures. Which is actually why I started looking for this issue... I'm having trouble figuring out where strapi keeps the model data that I want.

**Comment by maturanomx** [comment]
> The only cons that I have against TypeScript is the ease to contribute to the project. It's far easier to contribute to a project without TS or Flow. However, the pros are stronger (type-checking, code confidence, etc) and we should definitely move to TS or Flow.

Last time I saw, TS could do type checking with JS files for free, nothing needed, and the type annotations could be done with JSdoc syntax. Is a big advantage of been a superset of JS. So, TS could be introduced gradually and at certain point, not need it to use all the characteristics and syntax.

Said that... also is not a **need** for the project, just a "*nice to have*". Bee careful when evaluating the effort and the priority.

Just my 2c.

**Comment by TimMensch** [comment]
"Best practices" are well beyond "nice to have" IMO.

That said, I was asking for TypeScript _support_. Porting the project to TypeScript isn't required for _support._

**Comment by maturanomx** [comment]
> "Best practices" are well beyond "nice to have" IMO.

Agree. And you don't need TS to follow best practices 🤷‍♂️ ...

But well. You could add support without much effort... Strapi is always been welcome contributions 👍

**Comment by TimMensch** [comment]
TypeScript really _should_ be considered a best practice. 

I've written a lot about this [elsewhere](https://realmensch.org/2017/01/03/static-types-considered-helpful/). I even wrote about [what made me change my mind about dynamic types](https://www.quora.com/What-was-the-final-straw-that-made-you-stop-programming-with-JavaScript/answer/Tim-Mensch), which didn't happen until I'd been using dynamic languages (JavaScript and others) for more than a decade (I've been programming for over 30 years).

We're well off topic now; let's move this to the comments on my article or on Quora if you would like to debate the points.

**Comment by Aurelsicoko** [comment]
I created a card on Product Board, feel free to upvote and give me more insights (see https://portal.productboard.com/strapi/c/30-support-typescript).

**Comment by Laureian** [comment]
I'm also waiting for TS in strapi as this will improve my productivity and I think this will help new strapi developers to start with. If there will be someone who will start with this feature I will help him.

**Comment by mikebevz** [comment]
hi guys, any update on that? In 2019 Typescript has definitely won ;)

**Comment by maxbaluev1** [comment]
Typescript awesome

**Comment by Aurelsicoko** [comment]
I've recently talked with @alexandrebodin about supporting TS into a generated project. It might be easier than we think but there is one downside: it will add more build time. 

**There are two different things that we have to consider:**
- support TS into a generated project
- convert/write the core (or new feature) of Strapi with TS

About the first point, we thought it could be a new step during the project generation (strapi new my-project) when we will ask if you want to generate a project in TypeScript or Vanilla JS.

What are your thoughts about this?

**Comment by John0x** [comment]
@Aurelsicoko Well, writing the core in typescript definitely has a lot of advantages, but typings for proper typescript support would already be enough. At least from a user perspective.

**Comment by Aurelsicoko** [comment]
@John0x I know it has a lot of advantages to write the core in TS but it has some downsides as well. We have to prioritize our work and as you said from a user perspective being able to generate TS projects could be a good temporary solution 👍

**Comment by ScreamZ** [comment]
I think supporting generating code is already good. Nothing prevent user to write plugins in typescript and use the builded JS in strapi.

It's about writing the core in ts, which provide better security due to types :)

**TLDR;**

I think it's cool but yeah, not really critical.
One cool thing could be to have a ambient type file which help while writing plugins

**Comment by tolstenko** [comment]
As @TimMensch mentioned, the lack of typescript support is the only thing that keeps my team using lb4 instead of strapi. For me, the strapi permissioning flow is easier to maintain and evolve than lb4.

**Comment by NikolasMelui** [comment]
> As @TimMensch mentioned, the lack of typescript support is the only thing that keeps my team using lb4 instead of strapi. For me, the strapi permissioning flow is easier to maintain and evolve than lb4.

It seems to me that these are completely different products, but I also know a few commands that use loopback4 or nest only because of typescript. I am sure that if strapi gets support for typescript - it will get more popularity!

**Comment by yagobski** [comment]
Typescript support will kill nestJS and LB4

**Comment by TimMensch** [comment]
What a curious thing to say.

NestJS works with or without TypeScript. How does "support" for TypeScript harm a project, ever? Answer: *It can't.* I can't even imagine what rational thought process could lead someone to that conclusion. 

Loopback 4 can even be used with JavaScript, though I think you'd need to be really shortsighted to do so:

https://strongloop.com/strongblog/loopback4-javascript-experience/

**Comment by derrickmehaffy** [comment]
Locking this as the recent conversations are not providing useful discussion.

**Comment by lauriejim** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/whats-the-best-practice-to-develop-the-api-in-typescript/642/1

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/strapi-integration-with-other-typescript-aplications-in-a-monorepo/1104/2
